### PR TITLE
arch: arm: nrf: add hardware description for nrf kconfig files

### PR DESCRIPTION
**Pull request by jakub-uC**
Simplify Kconfig.nrfx for serial to use HAS_HW_NRF_UART0.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

### CODE REVIEW

**Comment by jarz-nordic** [comment]
Purpose of this PR is to allow to display and configure in Kconfig only these drivers that are available for particular SoC.

**Comment by anangl** [comment]
@jarz-nordic I think such explanation (why all these config entries are introduced) should be contained in the commit message. And the change in Kconfig.nrfx for serial driver should be mentioned as an example of usage of these entries, or moved to a separate commit.

**Review by anangl — CHANGES_REQUESTED** [review]

File: arch/arm/soc/nordic_nrf/Kconfig [hunk 1]
```diff
@@ -15,5 +15,7 @@ config SOC_FAMILY
 	string
 	default "nordic_nrf"
 
+gsource "arch/arm/soc/nordic_nrf/Kconfig.peripherals"
```
Comment (ulfalizer) [inline arch/arm/soc/nordic_nrf/Kconfig]: A plain `source` would do here. `gsource` works like an optional include (like `-include` in `make`) when you don't have any wildcards in the filename.

**Review by ulfalizer — CHANGES_REQUESTED** [review]

File: arch/arm/soc/nordic_nrf/Kconfig.peripherals [hunk 2]
```diff
@@ -0,0 +1,208 @@
+# Kconfig.peripherals - Nordic Semiconductor nRFx MCU line
+#
+# Copyright (c) 2018-2018 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/Kconfig.peripherals]: fixed

File: arch/arm/soc/nordic_nrf/Kconfig.peripherals [hunk 3]
```diff
@@ -0,0 +1,208 @@
+# Kconfig.peripherals - Nordic Semiconductor nRFx MCU line
```
[the remark this answers is not in the document]
Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/Kconfig.peripherals]: changed

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.series [hunk 4]
```diff
@@ -8,15 +8,13 @@
 config  SOC_SERIES_NRF52X
 	bool "Nordic Semiconductor nRF52 series MCU"
 	select CPU_CORTEX_M4
-	select CPU_HAS_FPU
```
[the remark this answers is not in the document]
Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.series]: Now there are 3 logical (I hope) commits.

File: arch/arm/soc/nordic_nrf/Kconfig (hunk 1 above)
Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/Kconfig]: fixed

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc [hunk 5]
```diff
@@ -5,13 +5,177 @@
...
+	select HAS_HW_NRF_TWIS1
+	select HAS_HW_NRF_UART0
+	select HAS_HW_NRF_UARTE0
+	select HAS_HW_NRF_WDT
 
 config SOC_NRF52840
 	depends on SOC_SERIES_NRF52X
 	bool
+	select CPU_HAS_FPU
+	select XIP
```
Comment (ioannisg) [inline arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: Why XIP is not selected in nRF52832?

**Review by ioannisg — CHANGES_REQUESTED** [review]

**Review by ulfalizer — APPROVED** [review]
`gsource` nit fixed.

File: arch/arm/soc/nordic_nrf/Kconfig.peripherals [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+config HAS_HW_NRF_UART0
+	bool
+
+config HAS_HW_NRF_UARTE0
+	bool
+
+config HAS_HW_NRF_UARTE1
+	bool
+
+config HAS_HW_NRF_USB
```
Comment (nordic-krch) [inline arch/arm/soc/nordic_nrf/Kconfig.peripherals]: should be USBD

File: arch/arm/soc/nordic_nrf/nrf51/Kconfig.soc [hunk 7]
```diff
@@ -11,11 +11,71 @@ depends on SOC_SERIES_NRF51X
 
 config SOC_NRF51822_QFAA
 	bool "NRF51822_QFAA"
+	select HAS_HW_NRF_CCM
```
Comment (nordic-krch) [inline arch/arm/soc/nordic_nrf/nrf51/Kconfig.soc]: GPIOTE missing

File: arch/arm/soc/nordic_nrf/nrf51/Kconfig.soc [hunk 8]
```diff
@@ -11,11 +11,71 @@ depends on SOC_SERIES_NRF51X
...
+	select HAS_HW_NRF_TIMER1
+	select HAS_HW_NRF_TIMER2
+	select HAS_HW_NRF_TWI0
+	select HAS_HW_NRF_TWI1
+	select HAS_HW_NRF_UART0
+	select HAS_HW_NRF_WDT
 
 config SOC_NRF51822_QFAB
 	bool "NRF51822_QFAB"
+	select HAS_HW_NRF_CCM
```
Comment (nordic-krch) [inline arch/arm/soc/nordic_nrf/nrf51/Kconfig.soc]: GPIOTE missing

Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/nrf51/Kconfig.soc]: fixed

File: arch/arm/soc/nordic_nrf/nrf51/Kconfig.soc (hunk 7 above)
Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/nrf51/Kconfig.soc]: fixed

File: arch/arm/soc/nordic_nrf/Kconfig.peripherals (hunk 6 above)
Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/Kconfig.peripherals]: fixed

File: arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc (hunk 5 above)
Comment (jarz-nordic) [reply arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc]: fixed
I was told that this is different XIP. I've created one more configuration parameter:
HAS_HW_NRF_XIP.

**Review by ioannisg — CHANGES_REQUESTED** [review]
Some minor requested changes

**Comment by erwango** [comment]
This is duplicate from information available in device tree.
So one will have to populate this, maintain SoC dtsi file and make sure that they are aligned (else I guess there will be some issues in build system)..
Not sure this will make things more simple.

**Comment by jarz-nordic** [comment]
Hi,
Yes, in some way this is duplicate information with DTS but it was done for the reason.
This is our interim solution to let Kconfig know what peripherals are available to display and configure only available hardware.

There is an ogoing discussion how to solve it in future:
https://github.com/zephyrproject-rtos/zephyr/issues/7302
but until that time we want to use proposed solution.
Our approach is simplar to proposals that are made, so we will need no much effort to adopt.

**Comment by jarz-nordic** [comment]
Fixed changes requested by @ioannisg

**Review by anangl — COMMENTED** [review]
Looks good.
But I think it would be more adequate to start the titles of the first two commits with "arch: nrf:" instead of "drivers: all:".

**Comment by erwango** [comment]
@jarz-nordic 

> This is our interim solution to let Kconfig ...
Thanks for explanation (it would have been useful to detail in initial PR description, but ok).

**Review by anangl — APPROVED** [review]

**Comment by ioannisg** [comment]
Looks good.
One last thing: "arch: nrf:" is wrong; we need "arch: arm: nrf: ..." instead.

**Comment by jarz-nordic** [comment]
I've changed prefix to: arch: arm: nrf:

**Review by ioannisg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by erwango** [comment]
My personal point of view is that Kconfig symbols to describe HW are duplicate information vs what should be extracted from dts. #7302 might eventually output to generation of some symbols that could then be used in Kconfig space.
Problem is that we're not even aligned on what output of #7302 should be.
I'm not sure everybody aligns on the need to be able to amend dts configuration by menuconfig.

**Comment by SebastianBoe** [comment]
This PR has the purpose of

> allow to display and configure in Kconfig only these drivers that are available for particular SoC.

It is reasonable to expect this from the configuration system. Duplicating information between DTS and Kconfig seems reasonable to me until #7302 is resolved.

**Comment by carlescufi** [comment]
@MaureenHelm Does the NXP hardware support use a similar approach to this?

**Comment by carlescufi** [comment]
@jarz-nordic could we make some of those common to all drivers instead of them being Nordic-specific?

**Comment by MaureenHelm** [comment]
> Does the NXP hardware support use a similar approach to this?

I've defined some HAS_* configs in `ext/hal/nxp/mcux/Kconfig` to set if a given peripheral is present in an SoC, and make the corresponding driver depend on that config. It's inconsistent though, because not every type of NXP peripheral has one of these configs, and thus there are cases that a user could incorrectly enable a driver for peripheral that doesn't exist in the SoC. I tried to reconcile this in #4317 but it was rejected on the grounds that I was introducing too many new configs.

**Comment by carlescufi** [comment]
@MaureenHelm thanks for the input. I spoke to @nashif and I think we need to reconsider this. It might take a while until we get a proper way to obtain DT info from Kconfig, so in the meantime we could have `HAS_HW_ADC` which would be shared by all hardware implementations, including NXP and Nordic. I suggest we put all of those in a centralized place instead of having them spread around the `drivers/` folder.
So something like `dts/common/Kconfig.peripherals` where we define all of those. This file would be removed once we reach an agreement on how to properly fix this.

**Comment by MaureenHelm** [comment]
There are a few cases in the Kinetis family where we have multiple variants of a given peripheral type (e.g., SPI, DSPI, LPSPI), so it isn't sufficient to define a generic peripheral Kconfig option like `HAS_HW_SPI`

**Comment by nashif** [comment]
> so it isn't sufficient to define a generic peripheral Kconfig option like HAS_HW_SPI

It should be ok to define HAS_HW_DSPI and HAS_HW_LPSPI which can be used by other vendors if they support them. I know some of those would be vendor specific and only one vendor might have them, but at least in this case it is consistent and safe for future support.

What I dislike is having those (many) HAS_HW_XXX defined for every vendor individually.

**Comment by nashif** [comment]
> So something like dts/common/Kconfig.peripherals

I would not put this in dts/ and make it appear as a DTS thing, DTS does nothing with those configs, at some point when move this decision to DTS we can make the change and remove them all hopefully.

Waiting to hear from the participants of the meeting last week where DTS in general was discussed @erwango , @galak , @tbursztyka

**Comment by jarz-nordic** [comment]
What is wrong about having such file by each vendor individually?

In my opinion one file with peripheral list containing all vendors and all possible peripheral variants will make developers work more difficult.

@carlescufi : generic list of peripherals like HAS_HW_ADC is not enough for us. Some SoCs have ADC other SAADC only, some have SPI other SPIM or both.  The best example is with nrf52840. 
It has possibility to configure 2 serial ports at time: UART0 and UART**E**1 or UART**E**0 and UART**E**1. It does not have UART1.

**Comment by nordic-krch** [comment]
I'm not sure if this all should be in common kconfig with the list of all possible options when in many cases names may be unique to vendor. It will generate long list and when vendor will be adding new peripheral it will require change to common file. Let say that there is LPUART which is low power UART, but other vendor has LEUART which is low energy uart. Should it use LPUART or create LEUART in common list. If LPUART is used it may confuse somebody because there will be kconfigs like:

config LEUART0_ENABLE
   depends on LPUART0

Additionally, we might want to not only put peripherals there but also specific features in those peripherals. Example: SPI in device X has max speed 4M and in device Y has 8M. I want to limit menu for those instances of SPI which do not support 8M. That i can to by creating flag like HAS_HW_SPI3_8M. Such flag to do belong to generic, common file for sure.

Summarizing, I think that each vendor should has own list of peripherals/features since those be used only in Kconfig specific to that vendor and nowhere else.

**Review by nashif — APPROVED** [review]

**Comment by carlescufi** [comment]
Merging but keeping track of this in #8481

**Pull request merged**